File: src/creational/prototype/typescript/classes/car-cache.ts
import { BigCar, SmallCar } from './cars';
import { Car } from './interfaces/car-protocol';

type Ids = typeof ids;
export type IdsKeys = keyof Ids;

export const ids = {
  SmallCar: 0,
  BigCar: 1,
};

export class CarsCache {
  private static readonly _cars: Record<number, Car> = {};

  static getCar(id: IdsKeys): Car {
    if (CarsCache._cars[ids[id]] === undefined) this.load(id);
    return CarsCache._cars[ids[id]].clone();
  }

  static load(id: IdsKeys): void {
    const loadDict = {
      SmallCar: this.loadSmallCar,
      BigCar: this.loadBigCar,
    };
    loadDict[id] && loadDict[id]();
  }

  static loadBigCar(): void {
    console.log('\nLoading BigCar');

    const bigCar = new BigCar();
    bigCar.id = ids.BigCar;
    CarsCache._cars[bigCar.id] = bigCar;

    console.log('finished loading\n');
  }

  static loadSmallCar(): void {
    console.log('\nLoading SmallCar');

    const smallCar = new SmallCar();
    smallCar.id = ids.SmallCar;

    CarsCache._cars[ids.SmallCar] = smallCar;

    console.log('finished loading\n');
  }
}

File: src/creational/prototype/python/classes/car_cache.py
from enum import Enum
from typing import Callable, Dict

from classes.cars import BigCar, SmallCar
from classes.interfaces.car_protocol import Car


class Ids(Enum):
    BIG_CAR = '1'
    SMALL_CAR = '2'


class CarsCache:
    __CARS: Dict[Ids, Car] = {}

    @staticmethod
    def __load(id: Ids) -> None:
        CAR_DICT: Dict[Ids, Callable] = {
            Ids.BIG_CAR: CarsCache.__load_big_car,
            Ids.SMALL_CAR: CarsCache.__load_small_car,
        }

        CAR_DICT[id]()

    @staticmethod
    def get_car(id: Ids) -> Car:
        if id not in CarsCache.__CARS.keys():
            CarsCache.__load(id)

        CAR = CarsCache.__CARS.get(id, None)
        return CAR.clone()

    @staticmethod
    def __load_big_car() -> None:
        print('\nloading BigCar')

        BIG_CAR = BigCar()
        BIG_CAR.set_id(Ids.BIG_CAR)
        CarsCache.__CARS[BIG_CAR.get_id()] = BIG_CAR

        print('finished loading\n')

    @staticmethod
    def __load_small_car() -> None:
        print('loading SmallCar')

        SMALL_CAR = SmallCar()
        SMALL_CAR.set_id(Ids.SMALL_CAR)
        CarsCache.__CARS[SMALL_CAR.get_id()] = SMALL_CAR

        print('finished loading\n')
